File: frontend/src/services/supplier.service.ts

import instance from "../config/axios.config";
import type { ApiResponse, ErrorApiResponse } from "../types/api.type";
import type {
    CreateSupplierRequest,
    UpdateSupplierRequest,
    SupplierResponse
} from '../types/supplier.types';

const URL_API = '/suppliers';

const getSuppliers = async (): Promise<ApiResponse<SupplierResponse[]>> => {
    try {
        const res = await instance.get<ApiResponse<SupplierResponse[]>>(URL_API);
        return res.data;
    } catch (error) {
        return error as ErrorApiResponse;
    }
};

const getSupplierById = async (id: number): Promise<ApiResponse<SupplierResponse>> => {
    try {
        const res = await instance.get<ApiResponse<SupplierResponse>>(`${URL_API}/${id}`);
        return res.data;
    } catch (error) {
        return error as ErrorApiResponse;
    }
};

const createSupplier = async (data: CreateSupplierRequest): Promise<ApiResponse<SupplierResponse>> => {
    try {
        const res = await instance.post<ApiResponse<SupplierResponse>>(URL_API, data, {
            headers: {
                'Content-Type': 'application/json'
            }
        });
        return res.data;
    } catch (error) {
        return error as ErrorApiResponse;
    }
};

const updateSupplier = async (id: number, data: UpdateSupplierRequest): Promise<ApiResponse<SupplierResponse>> => {
    try {
        data.id = id;
        const res = await instance.patch<ApiResponse<SupplierResponse>>(URL_API, data, {
            headers: {
                'Content-Type': 'application/json'
            }
        });
        return res.data;
    } catch (error) {
        return error as ErrorApiResponse;
    }
};

const deleteSupplier = async (id: number): Promise<ApiResponse<boolean>> => {
    try {
        const res = await instance.delete(`${URL_API}/${id}`);
        if (res.status === 204) {
            return { data: true, success: true, message: '' } as ApiResponse<boolean>;
        }
        return res.data;
    } catch (error) {
        return error as ErrorApiResponse;
    }
};

export {
    getSuppliers,
    getSupplierById,
    createSupplier,
    updateSupplier,
    deleteSupplier
};

File: blazor_web/Services/Supplier/SupplierService.cs
using System.Net.Http.Json;
using blazor_web.DTOs.Supplier;
using blazor_web.Models;
using System.Collections.Generic;

namespace blazor_web.Services.Supplier
{
    public class SupplierService : ISupplierService
    {
        private readonly HttpClient _httpClient;
        private const string URL_API = "suppliers";

        public SupplierService(HttpClient httpClient)
        {
            _httpClient = httpClient;
        }

        // --- GET Suppliers ---
        public async Task<ApiResponse<List<SupplierResponse>>> GetSuppliersAsync()
        {
            try
            {
                var response = await _httpClient.GetFromJsonAsync<ApiResponse<List<SupplierResponse>>>(URL_API);

                return response ?? new ApiResponse<List<SupplierResponse>> { Success = false, Message = "Phản hồi API rỗng hoặc không hợp lệ." };
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error fetching suppliers: {ex.Message}");
                return new ApiResponse<List<SupplierResponse>> { Success = false, Message = $"Vui lòng điền đúng thông tin" };
            }
        }

        // --- GET Supplier By ID ---
        public async Task<ApiResponse<SupplierResponse>> GetSupplierByIdAsync(int id)
        {
            try
            {
                var response = await _httpClient.GetFromJsonAsync<ApiResponse<SupplierResponse>>($"{URL_API}/{id}");
                return response ?? new ApiResponse<SupplierResponse> { Success = false, Message = "Phản hồi API rỗng." };
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error fetching supplier {id}: {ex.Message}");
                return new ApiResponse<SupplierResponse> { Success = false, Message = $"Vui lòng điền đúng thông tin" };
            }
        }

        // --- POST Create Supplier ---
        public async Task<ApiResponse<SupplierResponse>> CreateSupplierAsync(CreateSupplierRequest request)
        {
            try
            {
                var httpResponse = await _httpClient.PostAsJsonAsync(URL_API, request);

                if (!httpResponse.IsSuccessStatusCode)
                {
                    var errorContent = await httpResponse.Content.ReadFromJsonAsync<ApiResponse<SupplierResponse>>();
                    return errorContent ?? new ApiResponse<SupplierResponse> { Success = false, Message = $"Tạo nhà cung cấp thất bại. (HTTP {httpResponse.StatusCode})" };
                }

                var successContent = await httpResponse.Content.ReadFromJsonAsync<ApiResponse<SupplierResponse>>();
                return successContent ?? new ApiResponse<SupplierResponse> { Success = false, Message = "Tạo thành công nhưng dữ liệu trả về rỗng." };
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error creating supplier: {ex.Message}");
                return new ApiResponse<SupplierResponse> { Success = false, Message = $"Vui lòng điền đúng thông tin" };
            }
        }

        // --- PATCH Update Supplier ---
        public async Task<ApiResponse<SupplierResponse>> UpdateSupplierAsync(int id, UpdateSupplierRequest request)
        {
            try
            {
                request.Id = id;

                var httpResponse = await _httpClient.PatchAsJsonAsync(URL_API, request);

                if (!httpResponse.IsSuccessStatusCode)
                {
                    var errorContent = await httpResponse.Content.ReadFromJsonAsync<ApiResponse<SupplierResponse>>();
                    return errorContent ?? new ApiResponse<SupplierResponse> { Success = false, Message = $"Cập nhật nhà cung cấp thất bại. (HTTP {httpResponse.StatusCode})" };
                }

                var successContent = await httpResponse.Content.ReadFromJsonAsync<ApiResponse<SupplierResponse>>();
                return successContent ?? new ApiResponse<SupplierResponse> { Success = false, Message = "Cập nhật thành công nhưng dữ liệu trả về rỗng." };
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error updating supplier {id}: {ex.Message}");
                return new ApiResponse<SupplierResponse> { Success = false, Message = $"Vui lòng điền đúng thông tin" };
            }
        }

        // --- DELETE Supplier ---
        public async Task<ApiResponse<bool>> DeleteSupplierAsync(int id)
        {
            try
            {
                var httpResponse = await _httpClient.DeleteAsync($"{URL_API}/{id}");

                if (httpResponse.IsSuccessStatusCode)
                {
                    return new ApiResponse<bool> { Success = true, Data = true, Message = "Xóa thành công." };
                }

                var errorContent = await httpResponse.Content.ReadFromJsonAsync<ApiResponse<bool>>();
                return errorContent ?? new ApiResponse<bool> { Success = false, Message = $"Xóa nhà cung cấp thất bại. (HTTP {httpResponse.StatusCode})" };
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error deleting supplier {id}: {ex.Message}");
                return new ApiResponse<bool> { Success = false, Message = $"Vui lòng điền đúng thông tin" };
            }
        }
    }
}